smarts/core/sumo_road_network.py

```diff
@@ -794,6 +821,33 @@ def distance_between(self, start: Point, end: Point) -> float:
                     d += road.length
             return d
 
+        @lru_cache(maxsize=8)
+        def project_along(
+            self, start: Point, distance: float
+        ) -> Set[Tuple[RoadMap.Lane, float]]:
+            route_roads = set(self._roads)
+            for cand_start_lane, _ in self._map.nearest_lanes(start, 30.0, False):
+                if cand_start_lane.road in route_roads:
+                    break
+            else:
+                logging.warning("unable to find road on route near start point")
+                return None
+            started = False
+            for road in self._roads:
+                if not started:
+                    if road != cand_start_lane.road:
+                        continue
+                    started = True
+                    lane_pt = cand_start_lane.to_lane_coord(start)
+                    start_offset = lane_pt.s
+                else:
+                    start_offset = 0
+                if distance > road.length - start_offset:
+                    distance -= road.length - start_offset
+                    continue
+                return {(lane, distance) for lane in road.lanes}
```

Comment:
How does this handle if there are forking roads to project along? ex.

![image](https://user-images.githubusercontent.com/44624435/133816959-ee73b590-ba14-49c0-8fd6-d0dd64491dc3.png)

From my understanding, it will return lanes of one of the forked roads in the example above

Reply:
With a `Route` object, there won't be forks because the whole route is determined from start to end.

However you question is a good one for the _other_ version of this in the `Lane` class.  And my assumption there was that the call to `waypoint_paths_at_offset()` there would deal with this correctly, but now I think there's a bit more work to do.  `waypoint_paths_at_offset()` for a lane gives you paths for all of the places you could go _from this lane only_, but it doesn't consider what would happen if you changed lanes to another lane on the same road first.  So I need to add that in.

So thanks for asking this question!